mlir/lib/Catalyst/Transforms/DetectQNodes.cpp

```diff
@@ -0,0 +1,1272 @@
+bool callsMlirAsyncRuntimeIsValueError(LLVM::CallOp callOp);
+bool callsMlirAsyncRuntimeIsTokenError(Operation *possibleCall);
+bool callsMlirAsyncRuntimeIsValueError(Operation *possibleCall);
+bool hasAbortInBlock(Block *block);
+
+// Helper function for creating function declarations
+LLVM::LLVMFuncOp lookupOrCreatePersonality(ModuleOp moduleOp);
+LLVM::LLVMFuncOp lookupOrCreateAbort(ModuleOp moduleOp);
+LLVM::LLVMFuncOp lookupOrCreateMlirAsyncRuntimeSetValueError(ModuleOp moduleOp);
+LLVM::LLVMFuncOp lookupOrCreateMlirAsyncRuntimeSetTokenError(ModuleOp moduleOp);
+LLVM::LLVMFuncOp lookupOrCreateUnrecoverableError(ModuleOp moduleOp);
+LLVM::LLVMFuncOp lookupOrCreateAwaitTokenName(ModuleOp);
+LLVM::LLVMFuncOp lookupOrCreateAwaitValueName(ModuleOp);
+LLVM::LLVMFuncOp lookupOrCreateDropRef(ModuleOp);
+
+void collectValuesToLookFor(ResultRange &results, SmallVector<Value> &valuesToLookFor);
+void collectValuesToLookFor(ResultRange &results, SmallVector<Value> &tokens,
+                            SmallVector<Value> &values);
+void collectResultsForMlirAsyncRuntimeErrorFunctions(SmallVector<Value> &values,
+                                                     SmallVector<Value> &results);
+void collectPotentialConditions(SmallVector<Value> &values, SmallVector<Value> &conditions);
+void collectSuccessorBlocks(SmallVector<Value> &conditions, SmallVector<Block *> &aborts,
+                            SmallVector<Block *> &success);
+void collectCallsToAbortInBlocks(SmallVector<Block *> &blocks, SmallVector<LLVM::CallOp> &calls);
+void replaceCallsWithCallToTarget(SmallVector<LLVM::CallOp> &oldCallOps, LLVM::LLVMFuncOp target,
+                                  SmallVector<LLVM::CallOp> &newCalls, PatternRewriter &rewriter);
+void replaceTerminatorWithUnconditionalJumpToSuccessBlock(SmallVector<Block *> abortBlocks,
+                                                          SmallVector<Block *> successBlocks,
+                                                          PatternRewriter &rewriter);
+
+// Actual content
+std::tuple<Block *, Block *, Block *> getBlocks(LLVM::CallOp callOp, PatternRewriter &rewriter);
+void setPersonalityAttribute(LLVM::LLVMFuncOp callerOp, LLVM::LLVMFuncOp personality,
+                             PatternRewriter &rewriter);
+void transformCallToInvoke(LLVM::CallOp callOp, Block *successBlock, Block *failBlock,
+                           PatternRewriter &rewriter);
+std::tuple<std::vector<Value>, std::vector<Value>>
+collectRefCountedTokensAndValues(LLVM::LLVMFuncOp funcOp);
+void insertCallToMlirAsyncRuntimeErrorFunction(Value value, LLVM::LLVMFuncOp fnDecl,
+                                               Block *failBlock, PatternRewriter &rewriter);
+void insertErrorCalls(std::vector<Value> tokens, std::vector<Value> values, Block *failBlock,
+                      PatternRewriter &rewriter);
+void insertBranchFromFailToSuccessor(Block *fail, Block *success, PatternRewriter &rewriter);
+
```

Comment:
please consider moving these helper functions to a different file.

Reply:
Moved them in several PRs.